src/nccl_ofi_net.c

```diff
@@ -1009,6 +1009,61 @@ int nccl_net_ofi_domain_fini(nccl_net_ofi_domain_t *domain)
 }
 
 
+int nccl_net_ofi_domain_release_all(nccl_net_ofi_device_t *device)
+{
+	int ret, first_error = 0, domain_num;
+
+	assert(device != NULL);
+	nccl_net_ofi_domain_t *domain, *domain_tmp;
+	nccl_net_ofi_ep_t *ep, *ep_tmp;
+
+	domain_num = HASH_COUNT(device->domain_table);
+	assert(domain_num > 0);
+	HASH_ITER(hh, device->domain_table, domain, domain_tmp) {
+		/* For each domain, clean up its endpoints. */
+		nccl_net_ofi_mutex_lock(&domain->domain_lock);
+		if (HASH_COUNT(domain->endpoint_table) > 0) {
+			/* Ignore ref_cnt of endpoints, since only called during plugin fini. */
+			HASH_ITER(hh, domain->endpoint_table, ep, ep_tmp) {
+				HASH_DEL(domain->endpoint_table, ep);
+
+				if (OFI_UNLIKELY(ep->ref_cnt > 0)) {
```

Comment:
similar to above, this should be a function call into the endpoint class, and not mucking with ref counts directly from the device.